builtin/credential/aws-iam/path_login.go

```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return logical.ErrorResponse("Error getting configuration"), nil
+	}
+
+	if config.HeaderValue != "" {
+		ok, msg := ensureVaultHeaderValue(headers, parsedUrl, config.HeaderValue)
+		if !ok {
+			return logical.ErrorResponse(fmt.Sprintf("Error validating %s header: %s", magicVaultHeader, msg)), nil
+		}
+	}
+
+	endpoint := config.Endpoint
+	if endpoint == "" {
+		endpoint = "https://sts.amazonaws.com"
+	}
+
+	clientArn, err := submitCallerIdentityRequest(method, endpoint, parsedUrl, body, headers)
+	if err != nil {
+		return logical.ErrorResponse(fmt.Sprintf("Error making upstream request: %s", err)), nil
+	}
+	canonicalArn, principalName, err := parseIamArn(clientArn)
+	if err != nil {
+		return logical.ErrorResponse("Unrecognized IAM principal type"), nil
+	}
+
+	roleName := data.Get("role").(string)
+	if roleName == "" {
+		roleName = principalName
+	}
+
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return logical.ErrorResponse(fmt.Sprintf("entry for role %s not found", roleName)), nil
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
+		return logical.ErrorResponse(fmt.Sprintf("IAM Principal '%s' does not belong to the role '%s'", clientArn, roleName)), nil
+	}
+
+	resp := &logical.Response{
+		Auth: &logical.Auth{
+			Policies: roleEntry.Policies,
+			Metadata: map[string]string{
+				"client_arn":    clientArn,
+				"canonical_arn": canonicalArn,
+			},
+			InternalData: map[string]interface{}{
+				"role_name": roleName,
+			},
+			DisplayName: principalName,
+			LeaseOptions: logical.LeaseOptions{
+				Renewable: true,
+				TTL:       roleEntry.TTL,
+			},
+		},
+	}
+
+	shortestTTL := b.System().DefaultLeaseTTL()
+	if roleEntry.TTL > time.Duration(0) && roleEntry.TTL < shortestTTL {
+		shortestTTL = roleEntry.TTL
+	}
+
+	maxTTL := b.System().MaxLeaseTTL()
+	if roleEntry.MaxTTL > time.Duration(0) && roleEntry.MaxTTL < maxTTL {
+		maxTTL = roleEntry.MaxTTL
+	}
+
+	if shortestTTL > maxTTL {
+		resp.AddWarning(fmt.Sprintf("Effective TTL of %q exceeded the effective max_ttl of %q; TTL value is capped accordingly", (shortestTTL / time.Second).String(), (maxTTL / time.Second).String()))
+		shortestTTL = maxTTL
+	}
+
+	resp.Auth.TTL = shortestTTL
+
+	return resp, nil
+}
+
+func (b *backend) pathLoginRenew(
+	req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	canonicalArn := req.Auth.Metadata["canonical_arn"]
+	if canonicalArn == "" {
+		return nil, fmt.Errorf("unable to retrieve canonical ARN from metadata during renewal")
+	}
+
+	roleName := req.Auth.InternalData["role_name"].(string)
+	if roleName == "" {
+		return nil, fmt.Errorf("Error retrieving role_name during renewal")
+	}
+	roleEntry, err := b.lockedAWSRole(req.Storage, roleName)
+	if err != nil {
+		return nil, err
+	}
+	if roleEntry == nil {
+		return nil, fmt.Errorf("role entry not found")
+	}
+
+	if roleEntry.BoundIamPrincipal != canonicalArn {
```

Comment:
Heh, there are MORE than enough sections of the AWS documentation to confuse people on this topic ;)

Anyway, cool, I'll update the docs (and try to avoid confusing people less than the AWS docs do...)